transport-native-io_uring/src/main/c/netty_io_uring_native.c

```diff
@@ -0,0 +1,297 @@
+
+	return 0;
+err:
+	io_uring_unmap_rings(sq, cq);
+	return ret;
+}
+
+void setup_io_uring(int ring_fd, struct io_uring *io_uring_ring,
+                    struct io_uring_params *p) {
+    int ret;
+
+    ret = io_uring_mmap(ring_fd, p, &io_uring_ring->sq, &io_uring_ring->cq);
+    if (!ret) {
+        io_uring_ring->flags = p->flags;
+        io_uring_ring->ring_fd = ring_fd;
+    } else {
+        //Todo signal this back to EventLoop
+        perror("setup_io_uring error \n");
+    }
+}
+
+static jint netty_io_uring_enter(JNIEnv *env, jclass class1, jint ring_fd, jint to_submit,
+                                 jint min_complete, jint flags) {
+    return sys_io_uring_enter(ring_fd, to_submit, min_complete, flags, NULL);
+}
+
+static int nettyBlockingSocket(int domain, int type, int protocol) {
+    return socket(domain, type, protocol);
+}
+
+static jobject netty_io_uring_setup(JNIEnv *env, jclass class1, jint entries) {
+    struct io_uring_params p;
+    memset(&p, 0, sizeof(p));
+
+    int ring_fd = sys_io_uring_setup((int)entries, &p);
+
+    //Todo
+    if (ring_fd < -1) {
+      //throw Exception
+      return NULL;
+    }
+    struct io_uring io_uring_ring;
+    //Todo memset instead
+    io_uring_ring.flags = 0;
+    io_uring_ring.sq.sqe_tail = 0;
+    io_uring_ring.sq.sqe_head = 0;
+    setup_io_uring(ring_fd, &io_uring_ring, &p);
+
+    jobject ioUringSubmissionQueue = (*env)->NewObject(
+        env, ioUringSubmissionQueueClass, ioUringSubmissionQueueMethodId,
+        (jlong)io_uring_ring.sq.khead, (jlong)io_uring_ring.sq.ktail,
+        (jlong)io_uring_ring.sq.kring_mask,
+        (jlong)io_uring_ring.sq.kring_entries, (jlong)io_uring_ring.sq.kflags,
+        (jlong)io_uring_ring.sq.kdropped, (jlong)io_uring_ring.sq.array,
+        (jlong)io_uring_ring.sq.sqes, (jlong)io_uring_ring.sq.ring_sz,
+        (jlong)io_uring_ring.cq.ring_ptr, (jint)ring_fd);
+
+    jobject ioUringCompletionQueue = (*env)->NewObject(
+        env, ioUringCompletionQueueClass, ioUringCommpletionQueueMethodId,
+        (jlong)io_uring_ring.cq.khead, (jlong)io_uring_ring.cq.ktail,
+        (jlong)io_uring_ring.cq.kring_mask,
+        (jlong)io_uring_ring.cq.kring_entries,
+        (jlong)io_uring_ring.cq.koverflow, (jlong)io_uring_ring.cq.cqes,
+        (jlong)io_uring_ring.cq.ring_sz, (jlong)io_uring_ring.cq.ring_ptr,
+        (jint)ring_fd);
+
+    jobject ringBuffer =
+        (*env)->NewObject(env, ringBufferClass, ringBufferMethodId,
+                        ioUringSubmissionQueue, ioUringCompletionQueue);
+
+    return ringBuffer;
+}
+
+static jint netty_create_file(JNIEnv *env, jclass class) {
+    return open("io-uring-test.txt", O_RDWR | O_TRUNC | O_CREAT, 0644);
+}
+
+static void netty_io_uring_native_JNI_OnUnLoad(JNIEnv *env) {
+    // Todo OnUnLoad
+}
+
+// JNI Method Registration Table Begin
+static const JNINativeMethod method_table[] = {
+    {"ioUringSetup", "(I)Lio/netty/channel/uring/RingBuffer;", (void *)netty_io_uring_setup},
```

Comment:
nit: this needs some love to allow shading of netty. Please see how we do this in other JNI code in epoll.

Reply:
sorry I dont get it

Reply:
https://github.com/netty/netty/blob/4.1/transport-native-epoll/src/main/c/netty_epoll_linuxsocket.c#L713